lib/broadway_kafka/producer/buffer.ex

```diff
@@ -0,0 +1,105 @@
+defmodule BroadwayKafka.Producer.Buffer do
+  @moduledoc false
+
+  # Implemenet Ring buffer using :queue
+  #
+  # Each item in the queue contains {key, list}
+  # which key is {generation_id, topic, partition} and list is
+  # the full messages to be read for this topic/partition.
+  # 
+  # When dequeuing messages, `dequeue` pops out an item in the queue
+  # takes `count` amount of messages for that {key, list} and put the rest back
+  # into the end of the queue.
+  #
+  # This approach ensures a more equitable consumption of all partitions.
+
+  alias Broadway.Message
+
+  @type t :: :queue.queue()
+  @type generation_id :: non_neg_integer
+  @type topic :: String.t()
+  @type partition :: non_neg_integer
+  @type key :: {generation_id, topic, partition}
+
+  @spec new :: t
+  @doc false
+  def new do
+    :queue.new()
+  end
+
+  @spec enqueue_with_key(t, key, list :: [Message.t()], at :: :rear | :front) :: t
+  @doc false
+  def enqueue_with_key(buffer, key, list, at \\ :rear)
+
+  def enqueue_with_key(buffer, _key, [], _at) do
+    buffer
+  end
+
+  def enqueue_with_key(buffer, key, list, at) do
+    if at == :rear do
+      :queue.in({key, list}, buffer)
+    else
+      :queue.in_r({key, list}, buffer)
+    end
+  end
+
+  @spec empty?(t) :: boolean
+  @doc false
+  defdelegate empty?(buffer), to: :queue, as: :is_empty
+
+  @spec dequeue(t, count :: non_neg_integer) ::
+          {t, list({key :: key, items :: list(Message.t()), last_item :: Message.t()}),
```

Comment:
personally I don't like the return value of this function but it's done to avoid some common operation on the list.

* t: the changed buffer
* list({key :: key, items :: list(Message.t()), last_item :: Message.t()}):
  * key: the key that the items where popped out, it's needed to do some ack operation.
  * items: the actual message that is going to be proceeded. 
  * last_item: The producer always needs the offset of the last popped item in a given key
*  dequeued_count: the number of items that are popped is needed to do new demand calculation.



In the ideal world, I'd like this function to only return: 
```elixir
{t, list(Message.t())}
```

but then we have to traverse the list multiple times to get:
* Number of dequeued items
* list of unique keys in the items
* the last item in for each set of keys